### PR TITLE
fix: test cases on windows OS

### LINKED ISSUE #1280: Test cases failing on Windows OS
### Prerequisites

The following tests are failing on windows, this is probably because of different path nomenclatures on windows & mac.

image

- [ ] I have checked the existing issues to ensure this is not a duplicate.
- [ ] I am running the latest version of Tailcall.

### Steps to reproduce

1.
2.
3. ... (add as many steps as needed)

### Expected behavior

A description of what you expected to happen.

### Actual behavior

A description of what actually happened.

### Environment information:

- **Operating System**: (e.g., Ubuntu 20.04, macOS Big Sur, Windows 10)
- **Tailcall Version**: (e.g., v1.0.2)
- **Rust Version**: (e.g., `rustc 1.58.0`)
- **GraphQL Client/version** (e.g., `curl/8.1.2`)

### Additional context

Add any other context about the problem here, such as relevant logs, error messages, or stack traces.

### Logs

```
Paste any relevant logs here.
```

### PR DESCRIPTION
**Pull request by mehulmathur16**
**Summary:**  
/claim #1280
/fixes #1280

**Issue Reference(s):**  
Fixes #... _(Replace "..." with the issue number)_

**Build & Testing:**

- [ ] I ran `cargo test` successfully.
- [ ] I have run `./lint.sh --mode=fix` to fix all linting issues raised by `./lint.sh --mode=check`.

**Checklist:**

- [ ] I have added relevant unit & integration tests.
- [ ] I have updated the [documentation] accordingly.
- [ ] I have performed a self-review of my code.
- [ ] PR follows the naming convention of `<type>(<optional scope>): <title>`

[documentation]: https://github.com/tailcallhq/tailcallhq.github.io/tree/develop/docs

## Summary by CodeRabbit

- **New Features**
	- Enhanced file reading capabilities on Windows, ensuring smoother operation across different operating systems.
	- Improved error messaging in the app, with specific messages tailored to the operating system, making troubleshooting more intuitive for users encountering missing file errors.

### CODE REVIEW

File: src/config/reader.rs [hunk 1]
```diff
@@ -326,7 +330,17 @@ mod test_proto_config {
...
-            let actual = helper_map
+            let mut helper_map_clone = helper_map.clone();
+
+            if cfg!(windows) {
+                for item in &mut helper_map_clone {
+                    if let Some(name) = &mut item.name {
+                        *name = name.replace('/', "\\");
+                    }
+                }
+            }
```
Comment (tusharmath) [inline src/config/reader.rs:341]: Let's use a library to do this in a platform independent manner instead of handwriting the logic.

**Review by tusharmath — CHANGES_REQUESTED** [review]

**Comment by tusharmath** [comment]
@mehulmathur16 ru on this?

**Comment by mehulmathur16** [comment]
> @mehulmathur16 ru on this?

Yes @tusharmath . I have looked for path libraries, but was only able to find libraries having fixed paths for cache, local, home, etc. Not able to find some path_resolver libraries. Please suggest if you have any in mind.

**Comment by mehulmathur16** [comment]
@tusharmath can we continue with manual checks instead of external library ?

**Comment by tusharmath** [comment]
@mehulmathur16 Yes. Let's just move it to separate module that can then later be refactored out and replaced with a library in one single change instead of making that everywhere in the code base.

**Comment by tusharmath** [comment]
Is this ready @mehulmathur16?

File: src/config/reader.rs [hunk 2]
```diff
@@ -41,6 +41,10 @@ impl ConfigReader {
     async fn read_file<T: ToString>(&self, file: T) -> anyhow::Result<FileRead> {
         // Is an HTTP URL
         let content = if let Ok(url) = Url::parse(&file.to_string()) {
+            if cfg!(windows) && !url.scheme().starts_with("http") {
+                let content = self.runtime.file.read(&file.to_string()).await?;
+                return Ok(FileRead { content, path: file.to_string() });
+            }
```
Comment (ssddOnTop) [inline src/config/reader.rs:47]: why do we need this?

File: src/config/reader.rs [hunk 3]
```diff
@@ -339,7 +343,13 @@ mod test_proto_config {
...
         let reader = ConfigReader::init(runtime);
         let mut proto_no_pkg = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
-        proto_no_pkg.push("src/grpc/tests/proto_no_pkg.graphql");
+
+        if cfg!(windows) {
+            proto_no_pkg.push("src\\grpc\\tests\\proto_no_pkg.graphql");
+        } else {
+            proto_no_pkg.push("src/grpc/tests/proto_no_pkg.graphql");
+        }
+
```
Comment (ssddOnTop) [inline src/config/reader.rs:352]: Use PathBuf push pop instead

File: src/config/reader.rs [hunk 4]
```diff
@@ -348,15 +358,33 @@ mod test_proto_config {
...
         proto_no_pkg.push("news_no_pkg.proto");
-        let err = &validation.as_vec().first().unwrap().message;
+
+        if cfg!(windows) {
+            let err = &validation
+                .as_vec()
+                .first()
+                .unwrap()
+                .message
+                .replace('/', "\\\\");
```
Comment (ssddOnTop) [inline src/config/reader.rs:368]: what's the logic behind this?

File: src/grpc/protobuf.rs [hunk 5]
```diff
@@ -285,10 +285,17 @@ mod tests {
...
+            assert_eq!(
+                error.to_string(),
+                "The system cannot find the file specified. (os error 2)".to_string()
+            );
+        } else {
+            assert_eq!(
+                error.to_string(),
+                "No such file or directory (os error 2)".to_string()
+            );
+        }
```
[the remark this answers is not in the document]
Comment (ssddOnTop) [reply src/grpc/protobuf.rs:310]: Earlier it was used to assert some other kind of error, now we have some abstraction and it throws IO error. I think we can just keep it `assert!(foo.is_err())`

File: src/config/reader.rs (hunk 2 above)
Comment (mehulmathur16) [reply src/config/reader.rs:47]: Reqwest doesn't support reading local file paths on windows. 

https://github.com/seanmonstar/reqwest/issues/178

Comment (ssddOnTop) [reply src/config/reader.rs:47]: I don't think we support `file://` anyways.. this is redundant I think

File: src/config/reader.rs (hunk 4 above)
Comment (mehulmathur16) [reply src/config/reader.rs:368]: err contained a / in file paths, so replaced it with \\ for windows. I wasn't able to find the root cause of why / was present in the first place.

**Comment by tusharmath** [comment]
consider using https://github.com/chipsenkbeil/typed-path
@coderabbitai can you show an example of we can use the above library to make this PR simpler?

**Comment by tusharmath** [comment]
@mehulmathur16 I am closing this PR. Ideally we would like to use a library to handle this piece of code. Please reopen this PR when you have a simpler approach to solving this issue or feel free to pick any other issue in the mean time.

**Comment by mehulmathur16** [comment]
> @mehulmathur16 I am closing this PR. Ideally we would like to use a library to handle this piece of code. Please reopen this PR when you have a simpler approach to solving this issue or feel free to pick any other issue in the mean time.

Sure @tusharmath, also I think this will require a lot of changes in the code itself. As I had seen yesterday the implementation of link directive is also giving some issues in this [ plus some of the tests are generating new snapshots for windows OS]. Right now, we can make sure in the coming test related PRs code reviews, that only pathbuf is being used and paths are not hard-coded with forward slashes.

**Pull request closed without merging**